### PR TITLE
[hostcfgd] Fix race condition when systemd and hostcfgd act on same service at the same time

### PR DESCRIPTION
**Pull request by wenyiz2021**
- PR description:

This will fix issue: https://github.com/sonic-net/sonic-buildimage/issues/11665

1. Logic change:

This issue happens base on the knowledge that "After load_minigraph, systemd and hostcfgd could act together on 1 feature/service"

Systemd will start/stop service container after device bootup, before and after load_minigraph. While after load_minigraph, hostcfgd will re-sync feature state based on the logic defined in init_cfg.json.j2, these 2 flows have chance to acting on the same service at the same time, and leave service in a bad state, causing sanity check to fail.

This issue is now seen on multi-asic platform for dhcp_relay service, because:
Before load_minigraph, multi-asic platform has 'DEVICE_METADATA is undefined', and due to logic from PR: https://github.com/sonic-net/sonic-buildimage/pull/10474, `dhcp_relay` was `enabled`; While after load_minigraph, `dhcp_relay` feature status changed to `disabled`

Issue could be seen on other topologies if there are more services introduced, OR more services is changing their state during load_minigraph.

2. Add syslog info to help identify more info:
In hostcfgd there are 2 ways to call the update_systemd_config(), update_feature_state() functions.
One from host config deamon register_callback()
Another from  host config deamon start() and load() 
But it's unclear what the callers are when those functions were called with missing syslog info. 

- How did I do this?

1. When service/feature is starting but unstable, `systemctl status {service}` will say it is `activating`:
```
**/var/log/syslog:Aug 26 17:51:08.725436 vlab-08 INFO **systemd[1]: Starting dhcp_relay container...****
/var/log/syslog:Aug 26 17:51:08.811345 vlab-08 NOTICE root: Fast boot flag: dhcp_relay false.
/var/log/syslog:Aug 26 17:51:08.830713 vlab-08 NOTICE root: Warm boot flag: dhcp_relay false.
/var/log/syslog:Aug 26 17:51:08.839020 vlab-08 NOTICE root: Starting dhcp_relay service...

admin@vlab-08:~$ sudo systemctl status dhcp_relay
● dhcp_relay.service - dhcp_relay container
     Loaded: loaded (/lib/systemd/system/dhcp_relay.service; enabled; vendor preset: enabled)
    Drop-In: /etc/systemd/system/dhcp_relay.service.d
             └─auto_restart.conf
     **Active: activating (start-pre) since Fri 2022-08-26 17:51:08 UTC; 1s ago**
Cntrl PID: 24538 (dhcp_relay.sh)
      Tasks: 3 (limit: 8836)
     Memory: 8.6M
     CGroup: /system.slice/dhcp_relay.service
```
2. Similarly, When service/feature is stopping but unstable, `systemctl status {service}` will say it is `deactivating`:
```
**/var/log/syslog:Aug 26 17:50:36.481710 vlab-08 INFO **systemd[1]: Stopping dhcp_relay container.**..**
/var/log/syslog:Aug 26 17:50:36.571731 vlab-08 NOTICE root: Fast boot flag: dhcp_relay false.
/var/log/syslog:Aug 26 17:50:36.577520 vlab-08 NOTICE root: Warm boot flag: dhcp_relay false.
/var/log/syslog:Aug 26 17:50:36.583665 vlab-08 NOTICE root: Stopping dhcp_relay service...

admin@vlab-08:~$ sudo systemctl status dhcp_relay
● dhcp_relay.service - dhcp_relay container
     Loaded: loaded (/lib/systemd/system/dhcp_relay.service; enabled; vendor preset: enabled)
    Drop-In: /etc/systemd/system/dhcp_relay.service.d
             └─auto_restart.conf
     **Active: deactivating (stop) since Fri 2022-08-26 17:50:36 UTC; 775ms ago**
   Main PID: 7138 (dhcp_relay.sh); Control PID: 20214 (dhcp_relay.sh)
      Tasks: 6 (limit: 8836)
     Memory: 37.6M
     CGroup: /system.slice/dhcp_relay.service
             ├─ 7138 /bin/bash /usr/local/bin/dhcp_relay.sh wait
             ├─ 7151 /bin/bash /usr/bin/dhcp_relay.sh wait
             ├─ 7152 python3 /usr/local/bin/container wait dhcp_relay
             ├─20214 /bin/bash /usr/local/bin/dhcp_relay.sh stop
             ├─20322 /bin/bash /usr/bin/dhcp_relay.sh stop
             └─20326 python3 /usr/local/bin/container stop dhcp_relay
```
In hostcfgd, check and print current systemctl state, if service is not in stable state( `inactive` or `running`), warn out and sleep for 10 secs. 
Magic 10 sec comes from the worst case scenario that, on multi-asic vs, `systemd` takes <9 secs to start up service. While it takes <5 sec to stop the service, depends on VM's powerfullness. 10 sec is enough if hostcfgd acts `during` the starting by systemd, even in worst case.

- How did I verify results?

1. Before change:
```
**syslog:Aug 19 21:11:21.325303 vlab-08 INFO systemd[1]: Starting dhcp_relay container...**
syslog:Aug 19 21:11:22.120960 vlab-08 NOTICE root: Fast boot flag: dhcp_relay false.
syslog:Aug 19 21:11:22.140037 vlab-08 NOTICE root: Warm boot flag: dhcp_relay false.
syslog:Aug 19 21:11:22.195790 vlab-08 NOTICE root: Starting dhcp_relay service...
syslog:Aug 19 21:11:23.373633 vlab-08 INFO dhcp_relay.sh[15104]: Removing obsolete dhcp_relay container with HWSKU /usr/share/sonic/device/x86_64-kvm_x86_64_4_asic-r0
syslog:Aug 19 21:11:23.862720 vlab-08 INFO dhcp_relay.sh[15315]: dhcp_relay
syslog:Aug 19 21:11:23.866292 vlab-08 INFO dhcp_relay.sh[15104]: Creating new dhcp_relay container with HWSKU msft_four_asic_vs
syslog:Aug 19 21:11:24.670409 vlab-08 INFO dhcp_relay.sh[15402]: c1b15ef3663b24556ba8e961bb1ada9632bbb3945a44dde90d29c2db7cc4b07b
syslog:Aug 19 21:11:27.827704 vlab-08 INFO hostcfgd: Updating feature 'dhcp_relay' systemd config file related to auto-restart ...
syslog:Aug 19 21:11:27.828285 vlab-08 INFO hostcfgd: Feautre 'dhcp_relay' systemd config file related to auto-restart is updated!
syslog:Aug 19 21:11:28.377306 vlab-08 DEBUG container: read_data: config:True feature:dhcp_relay fields:[('set_owner', 'local'), ('no_fallback_to_local', False), ('state', 'disabled')] val:['local', False, "{% if not (DEVICE_METADATA is defined and DEVICE_METADATA['localhost'] is defined and DEVICE_METADATA['localhost']['type'] is defined and DEVICE_METADATA['localhost']['type'] is not in ['ToRRouter', 'EPMS', 'MgmtTsToR', 'MgmtToRRouter']) %}enabled{% else %}disabled{% endif %}"]
syslog:Aug 19 21:11:28.407568 vlab-08 DEBUG container: read_data: config:False feature:dhcp_relay fields:[('current_owner', 'none'), ('remote_state', 'none'), ('container_id', '')] val:['none', 'none', '']
syslog:Aug 19 21:11:28.408546 vlab-08 DEBUG container: container_start: dhcp_relay: set_owner:local fallback:True remote_state:none

**syslog:Aug 19 21:11:29.610500 vlab-08 INFO hostcfgd: Running cmd: 'sudo systemctl stop dhcp_relay.service'**
syslog:Aug 19 21:11:29.702286 vlab-08 INFO container: docker cmd: start for dhcp_relay
syslog:Aug 19 21:11:29.772835 vlab-08 NOTICE root: Started dhcp_relay service...
syslog:Aug 19 21:11:29.809068 vlab-08 WARNING systemd[1]: dhcp_relay.service: Control process exited, code=killed, status=15/TERM
**syslog:Aug 19 21:11:29.809217 vlab-08 INFO hostcfgd: Running cmd: 'sudo systemctl disable dhcp_relay.service'
syslog:Aug 19 21:11:29.809399 vlab-08 WARNING systemd[1]: dhcp_relay.service: Failed with result 'signal'.
syslog:Aug 19 21:11:29.809516 vlab-08 INFO systemd[1]: Stopped dhcp_relay container.**
syslog:Aug 19 21:11:29.610500 vlab-08 INFO hostcfgd: Running cmd: 'sudo systemctl stop dhcp_relay.service'
syslog:Aug 19 21:11:29.809217 vlab-08 INFO hostcfgd: Running cmd: 'sudo systemctl disable dhcp_relay.service'
syslog:Aug 19 21:11:31.282364 vlab-08 INFO hostcfgd: Running cmd: 'sudo systemctl mask dhcp_relay.service'


admin@vlab-08:/var/log$ sudo systemctl status dhcp_relay
Warning: The unit file, source configuration file or drop-ins of dhcp_relay.service changed on disk. Run 'systemctl daemon-reload' to reload units.
â— dhcp_relay.service
     Loaded: masked (Reason: Unit dhcp_relay.service is masked.)
    Drop-In: /etc/systemd/system/dhcp_relay.service.d
             â””â”€auto_restart.conf
     **Active: failed (Result: signal) since Fri 2022-08-19 21:11:29 UTC; 4min 13s ago**
```

2. After change, hostcfgd will catch unstable state and postpone to disable `dhcp_relay` service:
```
/var/log/syslog:Aug 26 21:31:19.309033 vlab-08 INFO **systemd[1]: Starting dhcp_relay container...**
/var/log/syslog:Aug 26 21:31:19.388668 vlab-08 NOTICE root: Fast boot flag: dhcp_relay false.
/var/log/syslog:Aug 26 21:31:19.405748 vlab-08 NOTICE root: Warm boot flag: dhcp_relay false.
/var/log/syslog:Aug 26 21:31:19.420069 vlab-08 NOTICE root: Starting dhcp_relay service...
/var/log/syslog:Aug 26 21:31:19.735838 vlab-08 INFO dhcp_relay.sh[31787]: Removing obsolete dhcp_relay container with HWSKU /usr/share/sonic/device/x86_64-kvm_x86_64_4_asic-r0
/var/log/syslog:Aug 26 21:31:19.962377 vlab-08 INFO dhcp_relay.sh[31914]: dhcp_relay
/var/log/syslog:Aug 26 21:31:19.964794 vlab-08 INFO dhcp_relay.sh[31787]: Creating new dhcp_relay container with HWSKU msft_four_asic_vs
/var/log/syslog:Aug 26 21:31:20.440955 vlab-08 INFO dhcp_relay.sh[32049]: 063bf7dffe5290ca2d229a8fa2ac7378c28649d840853f379abd08f49d0d0cf3
/var/log/syslog:Aug 26 21:31:21.743969 vlab-08 INFO hostcfgd: Updating feature 'dhcp_relay' systemd config file related to auto-restart ...
/var/log/syslog:Aug 26 21:31:21.744057 vlab-08 INFO hostcfgd: Feautre 'dhcp_relay' systemd config file related to auto-restart is updated!
**/var/log/syslog:Aug 26 21:31:22.592031 vlab-08 INFO hostcfgd: is_systemctl_status_stable: Active State for feature dhcp_relay is:      Active: activating (start-pre) since Fri 2022-08-26 21:31:19 UTC; 3s ago
/var/log/syslog:Aug 26 21:31:22.592144 vlab-08 WARNING hostcfgd: service dhcp_relay is transitioning, sleep 10 sec before disabling!**
/var/log/syslog:Aug 26 21:31:22.652692 vlab-08 DEBUG container: read_data: config:True feature:dhcp_relay fields:[('set_owner', 'local'), ('no_fallback_to_local', False), ('state', 'disabled')] val:['local', False, "{% if not (DEVICE_METADATA is defined and DEVICE_METADATA['localhost'] is defined and DEVICE_METADATA['localhost']['type'] is defined and DEVICE_METADATA['localhost']['type'] is not in ['ToRRouter', 'EPMS', 'MgmtTsToR', 'MgmtToRRouter']) %}enabled{% else %}disabled{% endif %}"]
/var/log/syslog:Aug 26 21:31:22.653308 vlab-08 DEBUG container: read_data: config:False feature:dhcp_relay fields:[('current_owner', 'none'), ('remote_state', 'none'), ('container_id', '')] val:['none', 'none', '']
/var/log/syslog:Aug 26 21:31:22.653571 vlab-08 DEBUG container: container_start: dhcp_relay: set_owner:local fallback:True remote_state:none
/var/log/syslog:Aug 26 21:31:23.230741 vlab-08 INFO container: docker cmd: start for dhcp_relay
/var/log/syslog:Aug 26 21:31:23.351539 vlab-08 NOTICE root: Started dhcp_relay service...
/var/log/syslog:Aug 26 21:31:23.383102 vlab-08 INFO systemd[1]: Started dhcp_relay container.

admin@vlab-08:~$ sudo systemctl status dhcp_relay
Warning: The unit file, source configuration file or drop-ins of dhcp_relay.service changed on disk. Run 'systemctl daemon-reload' to reload units.
● dhcp_relay.service
     Loaded: masked (Reason: Unit dhcp_relay.service is masked.)
    Drop-In: /etc/systemd/system/dhcp_relay.service.d
             └─auto_restart.conf
     **Active: inactive (dead)**
```

### CODE REVIEW

**Comment by yozhao101** [comment]
@wenyiz2021 Can you rephrase this PR title to describe what is the issue this PR aims to address please?

File: scripts/hostcfgd [hunk 1]
```diff
@@ -223,6 +224,8 @@ class FeatureHandler(object):
         # Enable/disable the container service if the feature state was changed from its previous state.
         if self._cached_config[feature_name].state != feature.state:
             if self.update_feature_state(feature):
+                syslog.syslog(syslog.LOG_INFO, "Status of feature '{}' is changed from '{}' to '{}' ..."
+                          .format(feature_name, self._cached_config[feature_name].state, feature.state))
```
Comment (yozhao101) [inline scripts/hostcfgd:228]: Indentation issue at line 228?

File: scripts/hostcfgd [hunk 2]
```diff
@@ -237,9 +240,10 @@ class FeatureHandler(object):
             if not feature_name:
                 syslog.syslog(syslog.LOG_WARNING, "Feature is None")
                 continue
-
+            syslog.syslog(syslog.LOG_INFO, "Re-sync the state field for feature:{} as the state field \
+                        might have been rendered based on DEVICE_METADATA table".format(feature_name))
```
Comment (yozhao101) [inline scripts/hostcfgd:244]: Indentation issue at line 244?

File: scripts/hostcfgd [hunk 3]
```diff
@@ -394,8 +398,9 @@ class FeatureHandler(object):
 
             for suffix in reversed(feature_suffixes):
                 cmds.append("sudo systemctl stop {}.{}".format(feature_name, suffix))
-                cmds.append("sudo systemctl disable {}.{}".format(feature_name, feature_suffixes[-1]))
-                cmds.append("sudo systemctl mask {}.{}".format(feature_name, feature_suffixes[-1]))
+            
+            cmds.append("sudo systemctl disable {}.{}".format(feature_name, feature_suffixes[-1]))
+            cmds.append("sudo systemctl mask {}.{}".format(feature_name, feature_suffixes[-1]))
```
Comment (yozhao101) [inline scripts/hostcfgd:403]: @wenyiz2021 I think the initial logic to disable a feature is to `stop`, `disable` the feature, then `mask` the correspoding systemd configuration file, right?

File: scripts/hostcfgd (hunk 1 above)
Comment (wenyiz2021) [reply scripts/hostcfgd:228]: @yozhao101 what should be the coding style here? thought spliting one line should have more indent on 2nd line

File: scripts/hostcfgd (hunk 2 above)
Comment (wenyiz2021) [reply scripts/hostcfgd:244]: @yozhao101 please suggest style here too

**Comment by wenyiz2021** [comment]
> @wenyiz2021 Can you rephrase this PR title to describe what is the issue this PR aims to address please?

thanks for pointing out! done

File: scripts/hostcfgd (hunk 1 above)
Comment (yozhao101) [reply scripts/hostcfgd:228]: Like this: 

https://github.com/sonic-net/sonic-host-services/blob/f9af7aee3468786d941ac38ebdd4c9654f559ef6/scripts/hostcfgd#L218

File: scripts/hostcfgd (hunk 3 above)
Comment (yxieca) [reply scripts/hostcfgd:403]: Are these 2 lines supposed to be in the for loop? If so, then you have indentation issue. Please set your edit to expand tabs to white spaces and use white spaces only.

**Comment by wenyiz2021** [comment]
++ @abdosi can you please help review this PR?
this is race condition that still happens today if one service was up and need to be brought down after load_minigraph.
although reboot could recover this, it's still a hidden issue.

**Comment by wenyiz2021** [comment]
dhcp_relay issue still occasionally happen and sometimes during test run, so not able to recover with reboot: 
E               Sep 21 12:18:53.347812 vlab-08 ERR monit[689]: 'container_checker' status failed (4) -- Unexpected running containers: dhcp_relay

https://dev.azure.com/mssonic/build/_build/results?buildId=151192&view=logs&j=54aa5ecf-394b-542e-d438-ceffe673928b&t=d5c1db74-40de-51fd-8e21-84b6f7edf3cc

This protection change is needed

File: scripts/hostcfgd [hunk 4]
```diff
@@ -353,10 +362,36 @@ class FeatureHandler(object):
...
         for feature_name in feature_names:
+            # Make sure service/feature state is stable before doing any change
+            # There is chance that the other flow, systemd will bring up/down the service
+            # systemd is acting slow, hostcfgd needs to wait till the service/feature become
+            # stable before enabling feature, ow feature will be left on a 'failed' state
+            if not self.is_systemctl_status_stable(feature_name):
+                syslog.syslog(syslog.LOG_WARNING, "service {} is transitioning, sleep 10 sec before disabling!"
+                              .format(feature_name))
+                # 10 seconds come from worst senario: on a multi-asic platform,
+                # where systemd start feature takes ~5sec, systemd stop takes ~8sec
```
Comment (SuvarnaMeenakshi) [inline scripts/hostcfgd:393]: the systemd time taken as 5sec and 8sec are observed on multi-asic virtual switch?

File: scripts/hostcfgd [hunk 5]
```diff
@@ -387,6 +422,16 @@ class FeatureHandler(object):
         cmds = []
         feature_names, feature_suffixes = self.get_multiasic_feature_instances(feature)
         for feature_name in feature_names:
+            # Make sure service/feature state is stable before doing any change
+            # There is chance that the other flow, systemd will bring up/down the service
+            # systemd is acting slow, hostcfgd needs to wait till the service/feature become
+            # stable before disabling feature, ow feature will be left on a 'failed' state
```
Comment (SuvarnaMeenakshi) [inline scripts/hostcfgd:428]: Typo - ow

File: scripts/hostcfgd [hunk 6]
```diff
@@ -402,10 +412,36 @@ class FeatureHandler(object):
...
+            True: if service status is stable
+            False: if service status is transitioning: Activating, Deactivating
+        """
+        cmd = "sudo systemctl status {}".format(service)
+        process = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        stdout, stderr = process.communicate()
+        for line in stdout.strip().split('\n'):
+            if "Active" in line:
+                syslog.syslog(syslog.LOG_INFO, "is_systemctl_status_stable: Active State for feature {} is: {}".format(service, line))
+                if "inactive (dead)" in line or "active (running)" in line:
```
Comment (SuvarnaMeenakshi) [inline scripts/hostcfgd:427]: There seems to be other states apart from inactive and active running.
Will it be better to just check if the service is transitioning (activating or deactivating) and return true upon that?

**Comment by SuvarnaMeenakshi** [comment]
@vaibhavhd / @arlakshm - in this PR, if there is a wait for 10seconds added, should there be any other checks or concerns that we have to keep in mind?
@vaibhavhd  - as discussed, can you add your suggestions here.

**Comment by wenyiz2021** [comment]
> @vaibhavhd / @arlakshm - in this PR, if there is a wait for 10seconds added, should there be any other checks or concerns that we have to keep in mind? @vaibhavhd - as discussed, can you add your suggestions here.

2 things I would like to clarify:
1. if it is a normal flow and service is not in transition state, this flow will act as old behavior, nothing changed
2. the issue now seen in multi-asic PR check, which has 64G vmss, is more powerful than my local 64G VM, so issue is less likely to happen, or even happens, the process is quicker, so I would say 10sec is a proper time.

File: scripts/hostcfgd (hunk 3 above)
Comment (yxieca) [reply scripts/hostcfgd]: My comments was a question, not a request. I was asking if your intended to put the 2 lines in the loop. Your answer is no. I think we are good.

Comment (wenyiz2021) [reply scripts/hostcfgd]: thanks. Sorry for the late reply, I forgot to send my review comments and just noticed it yday.

**Comment by wenyiz2021** [comment]
@SuvarnaMeenakshi @arlakshm @vaibhavhd Please kindly review/suggest

**Comment by yxieca** [comment]
@wenyiz2021 hostcfgd is not a submodule in 202205. If this fix is needed for 202205. please create a separate PR when you addressed all review issues in this PR.

**Review by kenneth-arista — APPROVED** [review]
Looks reasonable to me

**Comment by wenyiz2021** [comment]
closing PR as it doesn't work properly as tested
